monai/inferers/splitter.py

```diff
@@ -56,144 +57,210 @@ def __call__(self, inputs: Any) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
+    ) -> tuple[tuple[int, ...], tuple[float, ...] | tuple[int, ...], tuple[int, ...]]:
+        spatial_ndim = len(spatial_shape)
+        # patch_size
+        patch_size = ensure_tuple_rep(self.patch_size, spatial_ndim)
+        # overlap
+        overlap = ensure_tuple_rep(self.overlap, spatial_ndim)
+        overlap = tuple(o if p else type(overlap[0])(0) for o, p in zip(overlap, patch_size))
+        # offset
+        offset = ensure_tuple_rep(self.offset, spatial_ndim)
+        for off, ps, sh in zip(offset, patch_size, spatial_shape):
+            if off < -ps:
+                raise ValueError(f"Negative `offset` ({off}) cannot be larger than `patch_size` ({ps}) in magnitude.")
+            if off >= sh:
+                raise ValueError(f"`offset` ({off}) cannot be larger than inputs size ({sh}).")
+        return patch_size, overlap, offset
+
+    def _get_patch_tensor(
+        self, inputs: torch.Tensor, location: tuple[int, ...], patch_size: tuple[int, ...]
+    ) -> torch.Tensor:
+        slices = (slice(None),) * self.non_spatial_ndim + tuple(
+            slice(loc, loc + ps) for loc, ps in zip(location, patch_size)
+        )
+        patch = inputs[slices]
+        # send the patch to target device
+        patch.to(self.device)
+        return patch
+
+    def _get_patch_wsi(self, inputs: Any, location: tuple[int, int], patch_size: tuple[int, int]) -> torch.Tensor:
+        patch, _ = self.reader.get_data(wsi=inputs, location=location, size=patch_size)
+        # send the patch to target device
+        return ToTensor(device=self.device)(patch)
+
+    def __call__(self, inputs: torch.Tensor | PathLike) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
         """Split the input tensor into patches and return patches and locations.
 
         Args:
-            inputs: a torch.Tensor with BCHW[D] dimensions, representing an image or a batch of images.
+            inputs: either a torch.Tensor with BCHW[D] dimensions, representing an image or a batch of images,
+                or the file path to a whole slide image.
 
         Yields:
             tuple[torch.Tensor, Sequence[int]]: yields tuple of patch and location
         """
-        n_non_spatial_dims = 2
-        spatial_ndim = inputs.ndim - n_non_spatial_dims
-        spatial_shape = inputs.shape[n_non_spatial_dims:]
-        patch_size = self._get_valid_patch_size(spatial_ndim)
-        overlap = self._get_valid_overlap(spatial_ndim, patch_size)
-        offset = self._get_valid_offset(spatial_shape, spatial_ndim, patch_size)
+        spatial_shape: tuple
+        processed_input: Any
+        if isinstance(inputs, torch.Tensor):
+            processed_input = inputs
+            self.get_patch = self._get_patch_tensor
+            spatial_shape = inputs.shape[self.non_spatial_ndim :]
+        elif isinstance(inputs, (str, os.PathLike)):
+            self.get_patch = self._get_patch_wsi
+            processed_input = self.reader.read(inputs)
```

Comment:
If I understand correctly here, we can only read wsi from a path without any preprocessing(such as `ScaleIntensity`), right?

Reply:
You are right, the user cannot perform any transformation here, and the reason is WSI is usually not fit into memory so we don't have any transform in MONAI that acts on the entire WSI (if it doesn't fit into memory) and instead if acts upon patches, so this class should only do the splitting and the rest of transformation can be done in the `PatchInferer` where `preprocessing` argument can be any chain of transform on the patch.